### PR TITLE
Add missing mstatus.MPP legalization

### PR DESCRIPTION
**Pull request by ved-rivos**
`mstatus.MPP` legal values are User (`00) (if U-mode implemented), Supervisor (`01) (if S-mode implemented) and Machine (`11). Encoding `10 is illegal.

### CODE REVIEW

File: model/riscv_sys_regs.sail
Comment (arichardson) [inline model/riscv_sys_regs.sail:221]: Is there a specific reason this is legalized to User rather than Machine?

If not this could be simplified to always fall back to Machine if the value is not supported.

File: model/riscv_sys_regs.sail
Comment (arichardson) [inline model/riscv_sys_regs.sail:314]:
```suggestion
  let m = [m with MPP = legalize_priv_level(m[MPP])];
```

File: model/riscv_sys_regs.sail
Comment (arichardson) [reply model/riscv_sys_regs.sail:221]: Agreed choosing lower privilege makes sense - I'd just like to see a comment along the lines of "we are free to choose the legalized privileged, but choosing the lowest valid privilege level is the safest choice".

File: model/riscv_sys_regs.sail
Comment (ved-rivos) [reply model/riscv_sys_regs.sail:220]: The only supported combinations of privilege modes are M, M+U, M+S+U (Table 2; Priv) so configuring the model with M+S would illegal. Such an assertion may want to be placed in the ISA parser or knobs parser to not create a misconfigured model.

File: model/riscv_sys_regs.sail
Comment (ved-rivos) [reply model/riscv_sys_regs.sail:222]: For M+U, the likely simplest legalization would be have a single flop for privilege that is updated with the and of the two bits. Besides its a safer default. For M only, MPP would be hardwired to `11. For M+S+U the likely simplest legalization to update the MSB of the priv level to be the AND of the two bits. Besides it leads to a safer legal value.

File: model/riscv_sys_regs.sail
Comment (ved-rivos) [reply model/riscv_sys_regs.sail:314]: This is nicer. Had not seen it used elsewhere. Is this new supported syntax?

Comment (arichardson) [reply model/riscv_sys_regs.sail:314]: Yes I was also not aware of it until #373 but apparently it's been supported for quite a long time.

File: model/riscv_sys_regs.sail
Comment (arichardson) [inline model/riscv_sys_regs.sail:224]:
```suggestion
  /* 
   * When a mode value is not supported, we legalize it to the lowest
   * supported mode since that is the safest choice.
   * Note: the only valid combinations of supported modes are M, M+U, M+S+U.
   */
  match p {
    0b00 => if haveUsrMode() then p else privLevel_to_bits(Machine),
    0b01 => if haveSupMode() then p else privLevel_to_bits(User),
    0b10 => privLevel_to_bits(if haveUsrMode() then User else Supervisor),
    0b11 => p
  }
```
How about something like this, so we don't need to match a tuple?

Comment (ved-rivos) [reply model/riscv_sys_regs.sail:224]: For 0b10 - this code will lead to picking Supervisor on a M only machine... It could be fixed by - `0b10 => privLevel_to_bits(if haveUsrMode() then User else Machine)`

For 0b01 - this code will lead to picking User on a M only machine...it could be fixed by - `0b01 => if haveSupMode() then p else if haveUsrMode() then User else Machine,`

Comment (ved-rivos) [reply model/riscv_sys_regs.sail:224]: Yes this would work. Is this preferred? Tuples are avoided because they are slow?

Comment (arichardson) [reply model/riscv_sys_regs.sail:224]: I don't think that should matter - I just found the tuple version slightly more difficult to parse in my head. But that could just be a me problem :) 

I'll defer to @jrtc27 or @Alasdair for what the preferred style would be.

File: model/riscv_sys_regs.sail
Comment (ved-rivos) [reply model/riscv_sys_regs.sail:224]: added a comment to the function.

**Comment by nwf-msr** [comment]
Something also needs to happen at initialization, I think?  Right now https://github.com/riscv/sail-riscv/blob/d5e89a71e3a84495c1b88a7749c25fd6b9da684b/model/riscv_sys_control.sail#L553 leaves `mstatus[MPP]` as `0b00`, even if that's not a legal value (say, on a M-mode-only core).

**Comment by ved-rivos** [comment]
> Something also needs to happen at initialization, I think? Right now
> 
> https://github.com/riscv/sail-riscv/blob/d5e89a71e3a84495c1b88a7749c25fd6b9da684b/model/riscv_sys_control.sail#L553
> 
> leaves `mstatus[MPP]` as `0b00`, even if that's not a legal value (say, on a M-mode-only core).

Agree. This was missed. I have committed an update to fix this.

**Review by arichardson — APPROVED** [review]
LGTM. I slightly prefer using an `if/else` instead of tuple match (since the first element is only needed for one case), but happy either way.
```
  if not(haveUsrMode()) then privLevel_to_bits(Machine)
  else match p {
    0b01 => if haveSupMode() then p else privLevel_to_bits(User),
    0b10 => privLevel_to_bits(if haveSupMode() then Supervisor else Machine),
    _ => p
  }
```

**Comment by ved-rivos** [comment]
> We also came across this, so... alternative implementation, which I think is clearer:

This is nicer. Updated to this.

**Review by arichardson — APPROVED** [review]

**Review by Timmmm — APPROVED** [review]
LGTM, thanks!

**Comment by Timmmm** [comment]
@billmcspadden-riscv We should merge this fix IMO. It's been a few months and there are no issues with it.

**Comment by Timmmm** [comment]
I'll merge this next week if nobody objects.

**Pull request merged**
